### PR TITLE
Log `TensorBoard` histograms

### LINKED ISSUE #19743: Log `TensorBoard` histograms
### Description & Motivation

I'd like to be able to log histograms using `TensorBoard`. Looking at how multiple/single scalars are currently logged by `lightning.fabric.loggers.TensorBoardLogger` [1], I think `SummaryWriter.add_histogram` [2] could be called (non-empty `torch.Tensor` instances) in a similar way

Unfortunately, values passed into `LightningModule.log` are first type-checked by `LightningModule.__to_tensor` [3] - this enforces values to be either `dict` or single-value (0Dim) `torch.Tensor` instances. So in order to support histograms, this check would need to be loosen slightly...

  - [1] https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/loggers/tensorboard.py#L211
  - [2] https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter.add_histogram
  - [3] https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/module.py#L642

### Pitch

Firstly, I'm proposing [4] becomes:
```
if not (torch.numel(value) == 1 or (torch.numel(value) > 0 and value.ndim == 1)):
    raise ValueError(
        f"`self.log({name}, {value})` was called, but the tensor must have a single element,"
        f"or a single non-empty dimension. You can try doing `self.log({name}, {value}.mean())`"
    )
```

and [1] becomes (say):
```
try:
    if isinstance(v, dict):
        self.experiment.add_scalars(k, v, step)
    elif isinstance(v, Tensor):
        self.experiment.add_histogram(k, v, step)
    else:
        self.experiment.add_scalar(k, v, step)

# TODO(fabric): specify the possible exception
except Exception as ex:
    raise ValueError(
        f"\n you tried to log {v} which is currently not supported. Try a dict or a scalar/tensor."
    ) from ex
```

I'm sure you're keen not to modify `LightnginModule` too much, but I couldn't figure out a way to pass 1+Dim `Tensor` instances through to the `TensorBoardLogger` without loosing the type checking. 

- [4] https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/module.py#L648

### Alternatives

As a workaround, I'm currently using a subclass of `LightningModule` that overloads the `.log` method [5] with:
```
def log(self, name: str, value: _METRIC, *args, **kwargs):
    if isinstance(value, Tensor) and value.ndim > 0:
        for logger in self.loggers:
            if isinstance(logger, TensorBoardLogger):
                logger.experiment.add_histogram(name, value, self.global_step)

    else:
        super().log(name, value, *args, **kwargs)
```

Introducing `TensorBoard` specifics into the `LightningModule` class, is a horrible hack (hence wanting to incorporate the "nicer" fix above!)

- [5] https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/module.py#L371

### Additional context

I'm very happy to open a PR (I have one written up from exploring potential implementations) is the above sounds reasonable?

Many thanks!

### PR DESCRIPTION
**Pull request by dominicgkerr**
## What does this PR do?

Allows `LightningModule` to log `Tensorboard` histograms

Fixes #19743

<details>
  <summary><b>Before submitting</b></summary>

- [X] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by Borda — DISMISSED** [review]

**Review by SkafteNicki — CHANGES_REQUESTED** [review]
I do not think we should touch essentially the core logic of `self.log` to support one feature for one logger. Could we please instead have a specialized function for the tensorboard logger called `log_hist` similar to the specialized methods we have in other loggers:
https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.loggers.tensorboard.html#module-lightning.pytorch.loggers.tensorboard

cc: @Borda

**Comment by Borda** [comment]
> I do not think we should touch essentially the core logic of `self.log` to support one feature for one logger.

good call :)
